### ISSUE TITLE
How to use `memref.global` and `FlattenMemRefSubspanPass`?

### ISSUE THREAD

**Issue by benvanik** [google/iree#5764]
Constants are ending up in dispatch regions (which is great!):
```mlir
  %cst = constant dense<[[[[-1.000000e+00]], [[0.000000e+00]], [[1.000000e+00]]], [[[-2.000000e+00]], [[0.000000e+00]], [[2.000000e+00]]], [[[-1.000000e+00]], [[0.000000e+00]], [[1.000000e+00]]]]> : tensor<3x3x1x1xf32>
  %0 = memref.buffer_cast %cst : memref<3x3x1x1xf32>
  %13 = memref.subview %0[0, 0, 0, %arg2] [3, 3, 1, %12] [1, 1, 1, 1] : memref<3x3x1x1xf32> to memref<3x3x1x?xf32, affine_map<(d0, d1, d2, d3)[s0] -> (d0 * 3 + s0 + d1 + d2 + d3)>>
```

To get these lowering I need to use the `TensorConstantBufferizePass` and `FoldTensorExtractOpPass` to turn them into globals:
```mlir
memref.global "private" constant @__constant_3x3x1x1xf32 : memref<3x3x1x1xf32> = dense<[[[[-1.000000e+00]], [[0.000000e+00]], [[1.000000e+00]]], [[[-2.000000e+00]], [[0.000000e+00]], [[2.000000e+00]]], [[[-1.000000e+00]], [[0.000000e+00]], [[1.000000e+00]]]]>
func @foo() {
  %0 = memref.get_global @__constant_3x3x1x1xf32 : memref<3x3x1x1xf32>
  ...
  %21 = memref.load %0[%arg6, %arg7, %c0, %20] : memref<3x3x1x1xf32>
}
```

This would be fine for me, but unfortunately `FlattenMemRefSubspanPass` doesn't handle globals, so the memrefs don't get flattened to 1D as required.

Is there another mix of passes I can use to flatten this?
Or, how would I go about making the pass support these? Maybe just using some op (not sure what it is) to cast the global memref %0 value to `?` so that the linear index works?

**Comment by benvanik** [comment]
(my temporary workaround is `-iree-flow-inline-constants-max-byte-length=0` to avoid the constants, but I'd like to support them :)

**Comment by MaheshRavishankar** [comment]
You could just create a new symbol to flatten all the globals to a 1d memref. You will have to create a new attribute with the same values but different shape. I think the DenseElementsAttr allows you to iterate over all values of the constant (which could be expensive :P ).
Maybe it is possible to just grab the raw bytes and "move" them to the new constant. Dont have more idea of those off the top of my head.

**Comment by benvanik** [comment]
Ah so turn FlattenMemRefSubspanPass into a ModulePass (it's currently on functions), find the memref.globals and convert them to rank 1, then memref.get_global would get the 1D type, and the load should just work. That sounds like it may work!

**Comment by antiagainst** [comment]
Yeah. Right now the `FlattenMemRefSubspan` pass only recognizes `hal.interface.binding.subspan` ops as the source for memrefs (as the pass name suggests). To support `memref.global`, just need to add a pattern similar to [`FlattenBindingSubspan`](https://github.com/google/iree/blob/74e33026bc9972e31ee793205fdfada78f67fecd/iree/compiler/Conversion/Common/FlattenMemRefSubspanPass.cpp#L106-L125) to generate `memref.global` and` memref.get_global` with flattened memref.

Converting the shape of a DenseElementsAttr actually is simple: it has a `reshape` method directly to just change the shape metadata. One thing that might be a bit problematic is that right now the pass generates `memref<?xT>` (as that's how IREE treats those buffers really). What we have here is actually a "constant" memref and I'd suspect `memref.global` will need static shape. Could use a `memref.cast` to cast it back to `<?xT>` to see if it threads everything together.
